### DISCUSSION TITLE
Anyway to fix what U2NET removes?

### DISCUSSION THREAD

**Discussion by FASTANDEXTREME** [danielgatis/rembg#566] [Q&A]
Sometimes it removes objects that were not supposed to be removed. For example,
![experiment](https://github.com/danielgatis/rembg/assets/66107594/267194a9-9698-471c-873f-c41164a78ad9)
In this image, gun also got removed but it was part of the image and not background.

**Comment by jayfunc** [comment]
I was also struggling with such kind of requirement, I found these three repositories recently:

1. https://github.com/geekyutao/Inpaint-Anything
2. https://github.com/advimman/lama
3. https://github.com/facebookresearch/segment-anything

They provide implementation to achieve image segmentation and inpainting with mask. Sometimes it has a better result than rembg, sometimes it not.

**Comment by Renvect** [comment]
Been dealing with this for some time, and think we got solution for this but requires combo with OpenCV to perform dual analysis and creation of composite mask based on this analysis with couple more tricks. So far it successfully restores lost contours and details within. We are looking to enhance it and expand to variety of use cases. I will update soon with results after we run some additional tests as many struggle with this part.

**Comment by FASTANDEXTREME** [comment]
will be looking forward for it, please update if you have any progress.
Thank you.
